### PR TITLE
Fix testsuite for task storage

### PR DESCRIPTION
**Pull request by sirosen**
1. Ensure that it is possible to test against live S3
    (tox passenv settings + a pytest CLI option to set the bucket)
2. Rearrange dependencies so that we can test that the library is usable without boto3
3. Ensure untested lines are run by the testsuite (e.g. `StorageException.__str__`)
4. Improve error handling and testing around invalid `result_reference` data

With a couple of careully considered no-cover pragma comments, this now provides complete test coverage when the `--funcx-s3-bucket` argument is provided.
For example, I've been running under `tox -- --funcx-s3-bucket sirosen-funcx-test`

### CODE REVIEW

**Comment by sirosen** [comment]
[sc-11645]

File: src/funcx_common/task_storage/s3.py [hunk 1]
```diff
@@ -56,18 +56,40 @@ def _store_to_s3(self, task: TaskProtocol, result: str) -> None:
...
-                Key=task.result_reference["key"],
+            bucket = result_ref["s3bucket"]
+            key = result_ref["key"]
+        except KeyError as err:
+            raise StorageException(
+                "task {task.task_id} result_reference pointed to S3, but was missing "
+                "s3bucket or key"
+            ) from err
+
+        if not isinstance(bucket, str):
```
Comment (yadudoc) [inline src/funcx_common/task_storage/s3.py:73]: I don't think these type-checks add much since this is the class that is writing the result_ref.

Comment (sirosen) [reply src/funcx_common/task_storage/s3.py:73]: I don't see any justification for _not_ doing them.

Any code in the funcx services could modify a task and write that field. Especially if new versions of the storage class roll out with new behaviors, being pedantic about everything guards against unexpected misbehavior.

**Review by yadudoc — APPROVED** [review]
I have a minor comment. I see that that there are tests added that test for invalid references which the code now has explicit handling for. This looks good to go otherwise.

**Pull request merged**